### PR TITLE
Create asset and provisioning github workflow changes

### PR DESCRIPTION
**Pull request by anuragc617**

### CODE REVIEW

File: src/configure/clients/github/TemplateServiceClient.ts [hunk 1]
```diff
@@ -28,4 +28,23 @@ export class TemplateServiceClient {
...
+        return this.restClient.sendRequest2(
+            requestUri,
+            'POST',
+            '6.0-preview.1',
+            inputs
+        );
+    }
+
+    public async getTemplateFile(templateId: string, fileId: string): Promise<string> {
+        let requestUri = this.templateServiceUri + "/" + templateId + "/files" + fileId;
```
Comment (bishal-pdMSFT) [inline src/configure/clients/github/TemplateServiceClient.ts:43]: Does fileId contain a '/' at start? Should we put one here explicitly as well to ensure that url does not break

File: src/configure/configurers/configurerFactory.ts [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
 
 export class ConfigurerFactory {
-    public static GetConfigurer(sourceRepositoryDetails: GitRepositoryParameters, azureSession: AzureSession, subscriptionId: string): Configurer {
-        switch(sourceRepositoryDetails.repositoryProvider) {
+    public static GetConfigurer(sourceRepositoryDetails: GitRepositoryParameters, azureSession: AzureSession, subscriptionId: string, templateType: TemplateType, localGitRepoHelper: LocalGitRepoHelper): Configurer {
+        switch (sourceRepositoryDetails.repositoryProvider) {
             case RepositoryProvider.Github:
-                return extensionVariables.enableGitHubWorkflow ? new GitHubWorkflowConfigurer(azureSession, subscriptionId) : new AzurePipelineConfigurer(azureSession);
+                if (extensionVariables.enableGitHubWorkflow) {
+                    return templateType == TemplateType.LOCAL ? new GitHubWorkflowConfigurer(azureSession, subscriptionId) : new RemoteGitHubWorkflowConfigurer(azureSession, subscriptionId, localGitRepoHelper);
```
Comment (bishal-pdMSFT) [inline src/configure/configurers/configurerFactory.ts:15]: should we rename `GitHubWorkflowConfigurer` to `LocalGitHubWorkflowConfigurer`

File: src/configure/model/models.ts [hunk 3]
```diff
@@ -76,7 +76,6 @@ export class PipelineConfiguration {
     workingDirectory: string;
     params: { [key: string]: any } = {};
     assets: { [key: string]: any } = {};
-    parameters: { [key: string]: any } = {};
```
Comment (bishal-pdMSFT) [inline src/configure/model/models.ts:79]: what was the purpose of this property ?

File: src/configure/model/templateModels.ts [hunk 4]
```diff
@@ -13,18 +15,17 @@ export interface PipelineTemplate {
...
     templateId: string;
     workingDirectory: string;
     templateWeight: number;
     templateDescription: string;
     attributes: TemplateAttributes;
 }
 
-export interface RemotePipelineTemplate extends PipelineTemplate {
-    templateId: string;
+export interface RemotePipelineTemplate extends PipelineTemplate, ExtendedPipelineTemplate {
```
Comment (bishal-pdMSFT) [inline src/configure/model/templateModels.ts:26]: Just thinking loud about this. I am very clearly able to segregate what does this model represent. It is having three things:
1. ExtendedpipelineTemplate properties
2. Local pipeline template fills - templateType, targetKind etc
3. Working directory which is already part of templateInfo

I understand that we want to handle both local and remote template as same object and hence this class extends PipelineTemplate. So, basically this class represents all properties of extendedPipelineTemplate as well as some properties from local template.
But what I don't completely understand is why workingDirectory should be part of this. Working directory is already part of TemplateInfo. can we use that object?

File: src/configure/configure.ts [hunk 5]
```diff
@@ -450,7 +462,7 @@ class Orchestrator {
     }
 
     private async getSelectedPipeline(repoAnalysisResult: RepositoryAnalysisParameters): Promise<void> {
-        extensionVariables.templateServiceEnabled = false;
+        extensionVariables.templateServiceEnabled = true;
```
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:465]: please turn if false while merging PR

File: src/configure/configure.ts [hunk 6]
```diff
@@ -513,9 +525,12 @@ class Orchestrator {
         if (!repoAnalysisResult || !repoAnalysisResult.applicationSettingsList) {
             return;
         }
+        let workingDirectories = Array.from(new Set(this.inputs.potentialTemplates
+            .filter((template) => template.templateType === TemplateType.REMOTE)
+            .map((template: RemotePipelineTemplate) => template.workingDirectory)));
         var applicationSettings = repoAnalysisResult.applicationSettingsList.filter(applicationSetting => {
             if (this.inputs.pipelineConfiguration.template.templateType === TemplateType.REMOTE) {
-                return true;
+                return workingDirectories.indexOf(applicationSetting.settings.workingDirectory) > 0;
```
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:533]: `>= 0`

Comment (bishal-pdMSFT) [inline src/configure/configure.ts:533]: Would it be better to handle casing here. Although not required currently, it is better to not take dependency on such facts.

File: src/configure/utilities/templateConverter.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
+const templateParametersRegex = /(inputs|variables|assets|secrets)\.\w+/g;
```
Comment (bishal-pdMSFT) [inline src/configure/utilities/templateConverter.ts:1]: we also have `system`. Is that not being used in configuration mustache expressions?

File: src/configure/utilities/templateConverter.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+const templateEnclosedParametersRegex = /{{{?\s*(inputs|variables|assets|secrets)\.\w+\s*}?}}/g;
+const mustacheHelperRegex = /{?{{\s*#(\w+)(.*?)({?{{\s*\/\1)}}}?/;
```
Comment (bishal-pdMSFT) [inline src/configure/utilities/templateConverter.ts:3]: any reason why `{?` is applied to first `{` char and not third one like done one line above

Comment (bishal-pdMSFT) [inline src/configure/utilities/templateConverter.ts:3]: I could not understand this completely. Will catch you for this sometime.

File: src/configure/model/templateModels.ts (hunk 4 above)
Comment (anuragc617) [reply src/configure/model/templateModels.ts:26]: TemplateInfo is not being used anywhere. When we get the list of templateInfo, we directly store it in  RemotePipelineTemplate object and use it afterwards that's why workingDirectory is part of it.

File: src/configure/configure.ts (hunk 5 above)
Comment (anuragc617) [reply src/configure/configure.ts:465]: done

File: src/configure/configure.ts (hunk 6 above)
Comment (anuragc617) [reply src/configure/configure.ts:533]: done thanks for pointing it.

Comment (anuragc617) [reply src/configure/configure.ts:533]: done.

File: src/configure/utilities/templateConverter.ts (hunk 7 above)
Comment (anuragc617) [reply src/configure/utilities/templateConverter.ts:1]: I haven't seen it yet. I have added it in parameters regex.

File: src/configure/utilities/templateConverter.ts (hunk 8 above)
Comment (anuragc617) [reply src/configure/utilities/templateConverter.ts:3]: `{?` means optional, it is there to match parameters even if they are 2 or 3 braces.

File: src/configure/clients/github/TemplateServiceClient.ts (hunk 1 above)
Comment (anuragc617) [reply src/configure/clients/github/TemplateServiceClient.ts]: done

File: src/configure/model/models.ts (hunk 3 above)
Comment (anuragc617) [reply src/configure/model/models.ts:79]: it wasn't being used any where.

File: src/configure/utilities/templateConverter.ts (hunk 8 above)
Comment (anuragc617) [reply src/configure/utilities/templateConverter.ts]: sure.

File: src/configure/configurers/configurerFactory.ts (hunk 2 above)
Comment (anuragc617) [reply src/configure/configurers/configurerFactory.ts]: yes, done

**Comment by anuragc617** [comment]
> Please add proper telemetry for this flow.

already added telemetry.

**Pull request merged**